binders/kafka-binder/spring-cloud-stream-binder-kafka-streams/src/main/java/org/springframework/cloud/stream/binder/kafka/streams/SerdeResolverUtils.java

```diff
@@ -0,0 +1,241 @@
+/*
+ * Copyright 2022-2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.stream.binder.kafka.streams;
+
+import java.lang.reflect.Method;
+import java.lang.reflect.Type;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.stream.Collectors;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.kafka.common.serialization.Serde;
+import org.apache.kafka.common.serialization.Serdes;
+
+import org.springframework.beans.factory.annotation.AnnotatedBeanDefinition;
+import org.springframework.context.ConfigurableApplicationContext;
+import org.springframework.core.ResolvableType;
+import org.springframework.kafka.support.serializer.JsonSerde;
+import org.springframework.util.ClassUtils;
+
+/**
+ * Utility class that contains various methods to help resolve {@link Serde Serdes}.
+ *
+ * @author Chris Bono
+ * @since 4.0
+ */
+abstract class SerdeResolverUtils {
+
+	private static final Log LOG = LogFactory.getLog(SerdeResolverUtils.class);
+
+	/**
+	 *  Return the closest matching configured {@code Serde<?>} bean if one exists, or the specified
+	 *  {@code fallbackSerde}, or finally a standard default serde if no fallback specified.
+	 *
+	 * @param context the application context
+	 * @param targetType the target type to find the serde for
+	 * @param fallbackSerde the fallback serde in case no matching serde bean found in the context
+	 * @return serde to use for the target type
+	 */
+	static Serde<?> resolveForType(ConfigurableApplicationContext context, ResolvableType targetType, Serde<?> fallbackSerde) {
+
+		List<Serde<?>> matchingSerdes = findMatchingSerdes(context, targetType);
```

Comment:
I apologize in advance for the drastic split/refactor on a single commit. This class is largely the previous method in KeyValueResolver w/ the added "findMatching" API. It is largely unchanged in this method below this point.